libbpf-rs/src/iter.rs

```diff
@@ -0,0 +1,62 @@
+use nix::unistd::close;
+use nix::{errno, libc};
+use std::mem::{size_of, MaybeUninit};
+
+use crate::*;
+
+/// Represents a bpf iterator for reading kernel data structures. This requires
+/// Linux 5.8.
+pub struct Iter {
+    link: Link,
+}
+
+impl Iter {
+    pub fn new(link: Link) -> Self {
+        Self { link }
+    }
+
+    /// Read raw iterator as objects of type `T`.
+    ///
+    /// The data is read in chunks to an intermediate buffer of `BUF_SIZE`
+    /// elements of type `T`.
+    ///
+    /// # Safety
+    ///
+    /// The type layout of `T` must match the layout of the data structure being
+    /// dumped by the bpf program to guarantee that the returned objects are
+    /// valid.
+    pub unsafe fn read<T: Copy, const BUF_SIZE: usize>(&self) -> Result<Vec<T>> {
```

Comment:
Couple things:

* We want almost all APIs in libbpf-rs to be safe -- we'll generally make an API unsafe if it takes an unsafe parameter (eg pointer, FD)
* Taking in a type `T` and a constant `BUF_SIZE` seems unnecessarily complex. How about we do `pub fn read(&self, &mut [u8) -> Result<usize>`. No need to reinterpret bytes here -- we can let the user handle it just like w/ the map functions: https://docs.rs/libbpf-rs/0.7.2/libbpf_rs/struct.Map.html#method.lookup

Reply:
Makes sense -- leave it to the user to use `plain` or similar. Maybe we can get it to implement `std::io::Read`: https://doc.rust-lang.org/std/io/trait.Read.html#tymethod.read. That way the user can just use `read_to_end(&self, &mut Vec[i8])` to read into a vector rather than having to think about how what to set as size of the buffer. What do you think?

Reply:
Yeah that's a great idea. `impl Read for Link` sounds pretty ergonomic.

Reply:
Ok done. I've introduced an `OpenIter` so that the fd doesn't need to be closed in the read implementation (i.e. we don't have to read the entire file in one `read()` call, as expected by `std::io::Read`.